File: java-example/src/main/java/no/javatec/pubsubemulator/spock/pubsub/DemoListenerWithAck.java
package no.javatec.pubsubemulator.spock.pubsub;

import io.micronaut.context.annotation.Requires;
import io.micronaut.gcp.pubsub.annotation.MessageId;
import io.micronaut.gcp.pubsub.annotation.PubSubListener;
import io.micronaut.gcp.pubsub.annotation.Subscription;
import io.micronaut.messaging.Acknowledgement;
import no.javatec.pubsubemulator.spock.dto.SampleReturnMessage;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.concurrent.atomic.AtomicInteger;

import static no.javatec.pubsubemulator.spock.CustomEnvironment.PUBSUB_CONFIG;


@PubSubListener
@Requires(notEnv = {PUBSUB_CONFIG})
public class DemoListenerWithAck {

    private static final Logger log = LoggerFactory.getLogger(DemoListenerWithAck.class);

    private final AtomicInteger receiveCount = new AtomicInteger();

    public int getReceiveCount() {
        return receiveCount.get();
    }

    @SuppressWarnings("unused")
    @Subscription("${pubsub.subscription}")
    public void onMessageReceived(
            final SampleReturnMessage message,
            @MessageId final String messageId,
            final Acknowledgement acknowledgement) {

        receiveCount.incrementAndGet();

        log.info(getClass().getSimpleName() +
                " received message-id: " +
                messageId +
                " Message: " +
                message.returnMessage());

        acknowledgement.ack();
    }
}

File: kotlin-example/src/main/kotlin/no/javatec/pubsubemulator/kotest/pubsub/DemoListenerWithAck.kt
package no.javatec.pubsubemulator.kotest.pubsub

import io.micronaut.gcp.pubsub.annotation.MessageId
import io.micronaut.gcp.pubsub.annotation.PubSubListener
import io.micronaut.gcp.pubsub.annotation.Subscription
import io.micronaut.messaging.Acknowledgement
import no.javatec.pubsubemulator.kotest.dto.SampleReturnMessage
import no.javatec.pubsubemulator.kotest.loggerFor
import java.util.concurrent.atomic.AtomicInteger

@PubSubListener
class DemoListenerWithAck {

    private val log = loggerFor<DemoListenerWithAck>()
    val receiveCount: AtomicInteger = AtomicInteger(0)

    @Suppress("unused")
    @Subscription("\${pubsub.subscription}")
    fun onMessageReceived(
        message: SampleReturnMessage,
        @MessageId messageId: String,
        acknowledgement: Acknowledgement
    ) {
        receiveCount.incrementAndGet()

        log.info(
            javaClass.simpleName +
                    " received message-id: " +
                    messageId +
                    " Message: " +
                    message.returnMessage
        )

        acknowledgement.ack()
    }
}